### PR TITLE
Function to calculate crystal field energies and intensities

### LINKED ISSUE #15699: Crystal field: fitting function calculating peak positions and intensities.
Write a Mantid fitting function calculating energies (peak positions) and intensities for crystal field fitting.
This is a new type of fitting function in Mantid and some changes to the framework may be necessary.

This is a part of #15684

### PR DESCRIPTION
**Pull request by mantid-roman**
Added a fitting function doing crystal field calculations. It is a new type of function in Mantid. To make it work a new interface `IFunctionGeneral` was added.

**To test:**

Run this script. The output is a table workspace with two columns: one for the peak positions and one for the intensities.

```
    fun = 'name=CrystalFieldPeaks,Ion=Ce,Temperature=25.0,B20=0.37737,B22=3.9770,B40=-0.031787,B42=-0.11611,B44=-0.12544'
    EvaluateFunction(fun, None, OutputWorkspace='out')
```

Fixes #15699.

Release notes will be updated when umbrella issue #15684 is finished.

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the coding standards? Is it well structured with small focussed classes/methods/functions?
- [x] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [x] If there are changes in the release notes then do they describe the changes appropriately?
##### Functional Tests
- [x] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by FedeMPouzols** [comment]
I'm afraid this branch has developed a merge conflict with master.

**Comment by mducle** [comment]
@mantid-roman Thanks for this. I've checked the output table against McPhase and everything agrees. 

Note that the output peak intensities calculated here is in barns, whereas the reduced data (if normalised to a Vanadium standard) is in milibarns/steradian. Is it worth changing this calculation to also output in mb/sr (e.g. multiply the constant in line 534 of CrystalElectricField.cpp by 1000 and divide by 4pi)? 

Also, not relevant in this PR, but the calculated intensity here is the integrated intensity (area), rather than the peak height, so for the next step, if you can use peak functions which uses peak area as parameters rather than peak height, that would be better.

**Comment by mantid-roman** [comment]
@mducle I would prefer that this function returns only two columns: energy and intensity in predefined units (if you prefer mb/sr I can change). More flexible output can be achieved using custom python functions.

**Comment by mducle** [comment]
Ok. Let's leave it in barns, and put everything into the Python wrapper.

File: Framework/SINQ/test/PoldiPeakCollectionTest.h [hunk 1]
```diff
@@ -164,19 +164,21 @@ class PoldiPeakCollectionTest : public CxxTest::TestSuite {
 
     TS_ASSERT_EQUALS(collection.intensityType(), PoldiPeakCollection::Maximum);
 
-    TableWorkspace_sptr newDummy(m_dummyData->clone());
+    ITableWorkspace_sptr newDummy(m_dummyData->clone().release());
```
Comment (FedeMPouzols) [inline Framework/SINQ/test/PoldiPeakCollectionTest.h]: I think we're lucky and we don't need the `release()` call any longer after the boost version has been upgraded? As per #15921.

**Comment by FedeMPouzols** [comment]
@mducle , @mantid-roman : if you're happy with the current way this is working, I'd be happy to give it a quick look and test and put it for "shipping".

File: Framework/SINQ/test/PoldiPeakCollectionTest.h (hunk 1 above)
Comment (mantid-roman) [inline Framework/SINQ/test/PoldiPeakCollectionTest.h]: I'll merge master then.

**Comment by mducle** [comment]
I'm happy!

**Comment by martyngigg** [comment]
I focussed on the code review aspect and I'm happy for this to go. There are a few minor things that I think can be tidied up in maintenance rather than holding this up - https://github.com/mantidproject/mantid/issues/15956.

:shipit:

**Pull request merged**